File: server/internal/transactions/persister.go
package transactions

import (
	"context"
	"fmt"
)

type Persister struct {
	Store PersistStore
	// WithLLMStaging holds the LLM state stable through the upsert so disabling
	// the worker cannot clear rows before a concurrent sync stages them.
	WithLLMStaging func(upsert func(stageForLLM bool) error) error
}

func (p *Persister) Open(ctx context.Context) (chan<- PersistEvent, <-chan PersistResult) {
	events := make(chan PersistEvent)
	result := make(chan PersistResult, 1)
	go func() {
		var counts ItemCounts
		var firstErr error
		for ev := range events {
			if firstErr != nil {
				continue
			}
			if err := ctx.Err(); err != nil {
				firstErr = err
				continue
			}
			firstErr = p.apply(ctx, ev, &counts)
		}
		result <- PersistResult{Counts: counts, Err: firstErr}
	}()
	return events, result
}

func (p *Persister) apply(ctx context.Context, ev PersistEvent, counts *ItemCounts) error {
	switch {
	case ev.AccountUpsert != nil:
		if err := p.Store.UpsertSyncedAccount(ctx, *ev.AccountUpsert); err != nil {
			return err
		}
		counts.AccountsUpserted++
		return nil
	case ev.Upsert != nil:
		draft := *ev.Upsert
		var isNew bool
		upsert := func(stageForLLM bool) error {
			draft.StageForLLM = stageForLLM
			var err error
			isNew, err = p.Store.UpsertSyncedTransaction(ctx, draft)
			return err
		}
		var err error
		if p.WithLLMStaging != nil {
			err = p.WithLLMStaging(upsert)
		} else {
			err = upsert(false)
		}
		if err != nil {
			return err
		}
		if isNew {
			counts.Added++
		} else {
			counts.Modified++
		}
		return nil
	case ev.Removal != nil:
		if _, err := p.Store.DeleteSyncedTransaction(ctx, ev.Removal.Source, ev.Removal.ID); err != nil {
			return err
		}
		counts.Removed++
		return nil
	case ev.Recurring != nil:
		chargeID, err := p.Store.UpsertRecurringCharge(ctx, *ev.Recurring)
		if err != nil {
			return err
		}
		return p.Store.ReplaceChargeTxns(ctx, chargeID, ev.Recurring.TransactionIDs)
	case ev.MarkRecurring != nil:
		return p.Store.MarkRecurringFromStreams(ctx, ev.MarkRecurring.SourceID)
	default:
		return fmt.Errorf("empty persist event")
	}
}

File: server-rs/src/transactions/sync/persister.rs
use anyhow::Result;
use sqlx::{Sqlite, SqlitePool, Transaction};

use crate::{
    accounts::{UpsertAccount, store::upsert_account},
    database,
    transactions::{
        AccountDraft, ItemCounts,
        store::{recurring_write, sync},
    },
};

use super::{PersistEvent, llm_worker::LlmWorker};

pub struct Persister {
    pool: SqlitePool,
    llm: std::sync::Arc<LlmWorker>,
}

impl Persister {
    pub(super) fn new(pool: SqlitePool, llm: std::sync::Arc<LlmWorker>) -> Self {
        Self { pool, llm }
    }

    pub(crate) async fn persist(&self, events: Vec<PersistEvent>) -> Result<ItemCounts> {
        let categorizer = self.llm.staging_guard().await;
        let stage_for_llm = categorizer.is_some();
        database::with_tx(&self.pool, |transaction| {
            Box::pin(async move {
                let mut counts = ItemCounts::default();
                for event in events {
                    Self::apply(transaction, event, stage_for_llm, &mut counts).await?;
                }
                Ok(counts)
            })
        })
        .await
    }

    async fn apply(
        transaction: &mut Transaction<'_, Sqlite>,
        event: PersistEvent,
        stage_for_llm: bool,
        counts: &mut ItemCounts,
    ) -> Result<()> {
        match event {
            PersistEvent::AccountUpsert(draft) => {
                let account = account(draft);
                upsert_account(&mut **transaction, &account).await?;
                counts.accounts_upserted += 1;
            }
            PersistEvent::Upsert(mut draft) => {
                draft.stage_for_llm = stage_for_llm;
                if sync::upsert_synced_transaction(transaction, &draft).await? {
                    counts.added += 1;
                } else {
                    counts.modified += 1;
                }
            }
            PersistEvent::Removal(removal) => {
                sync::delete_synced_transaction(transaction, removal.source, &removal.external_id).await?;
                counts.removed += 1;
            }
            PersistEvent::Recurring(charge) => {
                let charge_id = recurring_write::upsert_recurring_charge(transaction, &charge).await?;
                recurring_write::replace_charge_transactions(transaction, charge_id, &charge.transaction_ids).await?;
            }
            PersistEvent::MarkRecurring { source_id } => {
                recurring_write::mark_recurring_from_streams(transaction, source_id).await?;
            }
        }
        Ok(())
    }
}

fn account(draft: AccountDraft) -> UpsertAccount {
    UpsertAccount {
        external_id: draft.external_id,
        connection_id: Some(draft.connection_id),
        owner_id: draft.owner_id,
        name: draft.name,
        account_type: draft.account_type,
        subtype: draft.subtype,
        mask: draft.mask,
        notes: None,
        closed: false,
        hidden: false,
        needs_review: draft.needs_review,
    }
}
